Implement JDBC based Presto driver

jeff303 opened the pull request:
Pulling out common Presto code into new presto-common driver (modeled after the relationship between, ex: `googleanalytics` and `google)`

Putting common QP/HoneySQL logic into the new (abstract) :presto-common driver

Updating :presto driver to extend from the new common driver and only adding HTTP/REST API related methods

Adding implementation of Presto JDBC driver, named :presto-jdbc, extending from :presto-common and :sql-jdbc

Adapting code from the existing Presto driver where appropriate

Adding new dependency-satisfied? implementation for :env-var, to allow for a plugin to require an env var be set, and making the Presto JDBC driver depend on that (specifically: `mb-enable-presto-jdbc-driver`)

jeff303 committed:
Implement JDBC based Presto driver

Pulling out common Presto code into new presto-common driver (modeled after the relationship between, ex: `googleanalytics` and `google)`

Putting common QP/HoneySQL logic into the new (abstract) :presto-common driver

Updating :presto driver to extend from the new common driver and only adding HTTP/REST API related methods

Adding implementation of Presto JDBC driver, named :presto-jdbc, extending from :presto-common and :sql-jdbc

Using com.facebook.presto/presto-jdbc as underlying JDBC driver dependency (since this is explicitly for Presto clusters, as opposed to Trino)

Adapting code from the existing Presto driver where appropriate

Adding new dependency-satisfied? implementation for :env-var, to allow for a plugin to require an env var be set, and making the Presto JDBC driver depend on that (specifically: `mb-enable-presto-jdbc-driver`)

Adding CircleCI configuration to run against the newer Presto (0.254) Docker image

camsaul commented:
Are you planning on making this the new `:presto` driver once you have everything working? I really don't want to maintain and test two separate Presto drivers going forward, or confuse our users by offering two separate options without making it clear which one they should use.

Let's just change the target to `master` and replace the existing `:presto` driver with this one going forward

dpsutton commented on modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj:
```diff
@@ -0,0 +1,129 @@
+  (try
+   (let [sql (sql-jdbc.describe-database/simple-select-probe-query driver schema table-name)]
+        ;; if the query completes without throwing an Exception, we can SELECT from this table
+        (jdbc/reducible-query {:connection conn} sql)
```
do you need to do any ssh-tunnel work at these call sites or when getting the details?

jeff303 replied:
The scope of this has sort of blown up so much that I was thinking of leaving tunnel support out of this round.  But yeah, if it gets added then that will have to be accounted for in many places.

jeff303 committed:
Fix import ordering

buremba commented:
Is this driver going to support Trino as well or is it out of scope?

jeff303 commented:
> Is this driver going to support Trino as well or is it out of scope?

It's out of scope for this change, unfortunately.  Trino is not client compatible with PrestoDB and the JDBC driver situation is a bit of a mess.  However, I'm hopeful that a lot of the work here can be reused if/when we do Trino (in the best case, this driver can serve as a parent and Trino can inherit the HoneySQL forms, etc. that would be equivalent between them).

jeff303 committed:
Fixing various test failures

Adding explicit ordering in a few places where it was missing

Fixing presto-type->base-type (timestamps were being synced as :type/Time because the regex pattern was wrong)

buremba commented:
@jeff303 I believe that that the dialect is identical so using the same driver with trino-jdbc would work just fine so the question is that if Metabase has plans to add support for Trino. :)

jeff303 commented:
> @jeff303 I believe that that the dialect is identical so using the same driver with trino-jdbc would work just fine so the question is that if Metabase has plans to add support for Trino. :) Would you consider merging this as something like `presto-base` and inherit `trino` and `presto` drivers from `presto-base` with just relevant jdbc drivers?

They're absolutely [not guaranteed to be compatible](https://trino.io/blog/2021/01/04/migrating-from-prestosql-to-trino.html#:~:text=Client%20protocol%20compatiblity), even if for now, they are.  At a minimum, the protocol is `jdbc:trino` as opposed to `jdbc:presto`, and given my experience in doing this work, I wouldn't be surprised at all if there are myriad more differences after that.  They forked a couple years ago and development has proceeded independently on each.

Even this patch is introducing a `presto-common` driver to be used between `presto` (the current Metabase Presto driver), and `presto-jdbc` (the one being added).  So a theoretical Trino driver could also inherit from this

jeff303 committed:
Fix ClassCastException

jeff303 committed:
Fix bug in timestamp sync type

Add tx/format-name test implementation for :presto-jdbc to lowercase table name

Make modified test Oracle friendly

jeff303 committed:
Fix bug in `if` form

jeff303 committed:
Fixing bug parsing the `[:zone :time]` case within `metabase.util.date-2.parse/parse-with-formatter`; the offset is nil so it can't be passed directly in this case, so use the `standard-offset` fn (which was moved from `date-2` to `common` to get a standard offset for that zone

Fixing more test failures by adding ordering

Changing sync to check whether the driver supports foreign keys before attempting to sync those (since drivers might throw an exception if attempting to check)

Moving some common test dataset functionality between :presto and :presto-jdbc to a new test.data ns for :presto-common

jeff303 committed:
Fix missing require

jeff303 committed:
Fix trailing blank line

Fix bug in new `standard-offset`, add test

jeff303 committed:
Fix bug in `[:zone :time]` clause for cases where the String has both the offset AND zone-id (ex: "13:30-07:00[America/Los_Angeles")

jeff303 committed:
Remove :schema from spec after handling

buremba commented:
@jeff303 you're right, my bad. Trino implements additional SQL syntaxes such as [MATCH_RECOGNIZE](https://trino.io/docs/current/sql/match-recognize.html) which is not available in Prestodb, which is the only change that I could find.

Trino will probably be adding more features over time but they use the same codebase that is being developed for 7 years at the beginning of 2020. I doubt Trino will change the syntax of the existing dialect since it will cause compatibility issues. That's why I assumed that this driver should just work for Trino. The primary reason for this change seems to be copyright issues, they're not trying to build a brand new SQL engine from scratch.

jeff303 committed:
Adding HoneySQL form for :absolute-datetime, to properly wrap in the from_iso8601_timestamp function (setting the value as a String literal in ISO 8601 zone offset format)

Adding HoneySQL form for :count-where, since we have to explicitly give a higher precision Decimal in order for Presto to not reduce the precision in results

Fixed a few tests

jeff303 committed:
Remove trace logging

Put limit within subquery for `expression-using-aggregation-test` (since ordering from subquery is not guaranteed in Presto)

jeff303 committed:
More timezone finagling, overriding some multimethods to make `time-timezone-handling-test` closer to passing

Reported issue with loss of millisecond portion to/from java.sql.Time: `https://github.com/dm3/clojure.java-time/issues/74`

jeff303 committed:
Add trace logging for sync statements

jeff303 committed:
Fixing timestamp handling for read-column-thunk implementations (omitting UTC Calendar param to the .get methods on ResultSet, since those seemed to actually be changing the TZ)

jeff303 committed:
Change trace to info level (temporarily)

jeff303 committed:
Fix import ordering and reflection warnings

jeff303 committed:
Add implementation switch to `read-column-thunk` for Types/TIMESTAMP to demonstrate different behaviors

jeff303 committed:
Still more timezone fixes

Adding impls for ->prepared-substitution to handle substitutions for native query params

Adding HoneySQL impls for `mod` (to do as "mod(x,y)" and `timestamp` (since it's a function with no parens to invoke) functions

Adding various `sql.qp/date` impls that use the `AT TIME ZONE` operator to account for report tz, and make bucketing tests happy

jeff303 committed:
Fix linter warnings

Fix reflection errors

jeff303 committed:
Fix tag for metabase/presto-mb-ci image

jeff303 committed:
Change ->honeysql impl for ZonedDateTime to use from_iso8601_timestamp fn (so arg can be parameterized)

Add MB_ENABLE_PRESTO_JDBC_DRIVER=true to CircleCI config

jeff303 committed:
More tweaking of timezone handling in Presto JDBC driver

Adding new dynamic var to query processor: *parent-expr-type-info*, to capture the type-info of the parent expression (in case the child needs to change its ->honeysql behavior according to the type of the expression, but it's not known at that level (because it's inside a `call`, for instance)

jeff303 committed:
Fix docstring/namespace/reflection errors

jeff303 committed:
Comment out unused fn

jeff303 committed:
Cleanup commit

Removing dead code

Reorganizing and adding section comments

Adding some tests

jeff303 committed:
Adjust batch sizes in load-data more, to make cast-test happy

jeff303 committed:
orders -> 30

jeff303 committed:
Try bumping up memory for presto latest to 2GB

jeff303 committed:
Switch back to 1GB

jeff303 committed:
Tweak CircleCI to capture max memory usage

jeff303 committed:
Adjust JAVA_TOOL_OPTIONS from 1GB to 512MB for presto-254

jeff303 committed:
switch to resource_class: large for presto-254

jeff303 committed:
Fix tx/dbdef->connection-details impl so it doesn't reimplement a bunch of logic

camsaul commented on .circleci/config.yml:
```diff
@@ -1073,12 +1101,57 @@ workflows:
+                command: cat /sys/fs/cgroup/memory/memory.max_usage_in_bytes
+                when: always
+          driver: presto-jdbc
+          extra-env: >-
+            MB_PRESTO_JDBC_TEST_CATALOG=test_data
+            MB_PRESTO_JDBC_TEST_HOST=localhost
+            MB_PRESTO_JDBC_TEST_PORT=8443
+            MB_PRESTO_JDBC_TEST_SSL=true
+            MB_PRESTO_JDBC_TEST_USER=metabase
+            MB_PRESTO_JDBC_TEST_PASSWORD=metabase
+            MB_ENABLE_PRESTO_JDBC_DRIVER=true
+            MB_PRESTO_JDBC_TEST_ADDITIONAL_OPTIONS="SSLTrustStorePath=/tmp/cacerts-with-presto-ssl.jks&SSLTrustStorePassword=changeit"
```
for the other drivers we just put the env vars in the executor themselves, and you don't need to add an `extra-env` parameter to do that. Guess it doesn't make a big difference tho. Doing the same thing 2 different ways doesn't do a lot to help the maintainability of the CircleCI config, which is already pretty hairy TBH

camsaul commented on enterprise/backend/test/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions_test.clj:
```diff
@@ -271,10 +271,15 @@
+        (is (= exp
+               (mt/rows
+                 (mt/with-gtaps {:gtaps {:venues (venue-names-native-gtap-def)}}
+                   (mt/run-mbql-query venues {:limit 2, :order-by [[:asc [:field (mt/id :venues :id)]]]})))))))
```
Use `mt/formatted-rows [int str]` instead of `mt/rows` and you won't need to add special casing for `:oracle`. It's better not to do that sort of special casing in the expected form if you can avoid it since it adds up quickly when other quirks come in to play

camsaul commented on enterprise/backend/test/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions_test.clj:
```diff
@@ -330,7 +335,8 @@
   (cond-> (mt/normal-drivers-with-feature :nested-queries :foreign-keys)
-    (@tx.env/test-drivers :bigquery) (conj :bigquery)))
+    (@tx.env/test-drivers :bigquery) (conj :bigquery)
+    :always                          (disj :presto-jdbc)))
```
`:always` is a bit weird here, everywhere else in the codebase we do this sort of thing we use `true`. I don't think there's an agreed-upon idiomatic key to use here like there is for `cond`, but we can at least stick to a common project-wide idiom

jeff303 replied:
I think I originally got it from [here](https://clojuredocs.org/clojure.core/cond-%3E#example-570c2e57e4b0fc95a97eab4a), but sure, I'll change to `true`.

camsaul commented on enterprise/backend/test/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions_test.clj:
```diff
@@ -915,7 +921,10 @@
+  (mt/test-drivers (disj
+                     (mt/normal-drivers-with-feature :foreign-keys :nested-queries :left-join)
+                     :redshift
+                     :presto-jdbc)
```
You should make a note of why we're not running this test for `:presto-jdbc`, similar to the note about why we're not running it for Redshift above

camsaul commented on modules/drivers/presto-common/src/metabase/driver/presto_common.clj:
```diff
@@ -0,0 +1,203 @@
+
+(defmethod sql.qp/date [:presto-common :day-of-week]
+  [_ _ expr]
+  (sql.qp/adjust-day-of-week :presto-common (hsql/call :day_of_week expr)))
```
You shouldn't hardcode `:presto-common` as the parameter here... it will mean any `:presto` or `:presto-jdbc` specific overrides for any methods that `sql.qp/adjust-day-of-week` ends up calling don't get used. Pass `driver` instead

camsaul commented on modules/drivers/presto-common/src/metabase/driver/presto_common.clj:
```diff
@@ -0,0 +1,203 @@
+
+(defmethod sql.qp/date [:presto-common :week]
+  [_ _ expr]
+  (sql.qp/adjust-start-of-week :presto-common (partial hsql/call :date_trunc (hx/literal :week)) expr))
```
Same thing goes here for my previous comment. Always pass `driver`, not the literal driver keyword, to other driver multimethods, so drivers that inherit from your driver work correctly

camsaul commented on modules/drivers/presto-common/test/metabase/driver/presto_common_test.clj:
```diff
@@ -0,0 +1,24 @@
+              [schema.core :as s]
+              [toucan.db :as db]))
+
+(use-fixtures :once (fixtures/initialize :db))
```
You can probably delete this namespace

camsaul commented on modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj:
```diff
@@ -0,0 +1,470 @@
+(defmethod sql.qp/unix-timestamp->honeysql [:presto-jdbc :microseconds]
+  [_ _ expr]
+  ;; Presto can't even represent microseconds, so convert to millis and call that version
+  (sql.qp/unix-timestamp->honeysql :presto-jdbc :milliseconds (hsql/call :/ expr 1000)))
```
Same thing applies here, you should pass `driver` rather than passing `:presto-jdbc` literally. You never know if we might want to use `:presto-jdbc` as the parent of something else in the future. (e.g. maybe in a test we want to mock out some situations so we create `::presto-jdbc-test`). It doesn't matter now since we have no child drivers, but it's better practice to write the code so it doesn't break if we add them

camsaul commented on modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj:
```diff
@@ -0,0 +1,470 @@
+                               ResultSet/TYPE_FORWARD_ONLY
+                               ResultSet/CONCUR_READ_ONLY)]
+       (try
+         (try
+           (.setFetchDirection stmt ResultSet/FETCH_FORWARD)
+           (catch Throwable e
+             (log/debug e (trs "Error setting statement fetch direction to FETCH_FORWARD"))))
+         stmt
```
I'm not sure this outer `try` is really needed. (I know this was copied this from the old `:presto` driver code, but now that I look at it again I don't think it makes sense, so better to fix past mistakes than to perpetuate them I think.) There's no possible situation where that the body is going to throw an Exception since the only thing we do is `.setFetchDirection` and that's already wrapped in its own try-catch.

camsaul commented on modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj:
```diff
@@ -0,0 +1,470 @@
+  [_ prepared-statement i ^OffsetDateTime t]
+  ;; necessary because `PrestoPreparedStatement` does not support `ZonedDateTime` in its setObject method
+  ;; convert to a SQL `timestamp` at UTC, and set that
+  (jdbc/set-parameter (t/sql-timestamp (t/with-zone-same-instant t (t/zone-id "UTC"))) prepared-statement i))
```
Is this easier than just converting the temporal objects to temporal literal strings like the stuff we looked at together a couple weeks ago? I'm wary about using `java.sql.Timestamp` since it does weird stuff. I guess it's fine if it works tho

jeff303 replied:
OK, I realized that what was needed was to override the `->honeysql` method for `OffsetDateTime` (it was already being done for `ZonedDateTime`).  I have made that change now, and this path is no longer exercised (and the tests pass).

However, is it still a good idea to override these?  If we allow it to fall back to the legacy impl, then the Presto JDBC driver will definitely throw a `NotImplementedException` (since the legacy method will invoke the setTimestamp method taking a `Calendar` as its last parameter, which the Presto JDBC driver does not implement).  So if it's possible there are any code paths that might still attempt to set this parameter (of type `OffsetDateTime`) then we may want to leave it in so they succeed instead of failing?

camsaul commented on modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj:
```diff
@@ -0,0 +1,470 @@
+  (let [report-zone       (qp.timezone/report-timezone-id-if-supported :presto-jdbc)
+        ^ZonedDateTime ts (if (str/blank? report-zone) t (t/with-zone-same-instant t (t/zone-id report-zone)))]
+    ;; the `from_iso8601_timestamp` only accepts timestamps with an offset (not a zone ID), so only format with offset
+    (date-time->substitution (.format ts DateTimeFormatter/ISO_OFFSET_DATE_TIME))))
```
You can probably use `(u.date/format t)` to do this if you don't want to deal with the Java methods. Doesn't really make a big difference tho I guess

Are you sure

camsaul replied:
It seems like having this output `timestamp '2021-06-23 14:09:00 -07:00'` or whatever the actual Presto literal format is might be preferable to `from_iso8601_timestamp(?)` but I guess it ultimately doesn't make a difference

jeff303 replied:
Yeah I changed these to use the first form (which is actually represented as a `CAST` in order to work with prepared statements/String param value).

camsaul commented on modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj:
```diff
@@ -0,0 +1,470 @@
+  [_ ^ResultSet rs ^ResultSetMetaData rs-meta ^Integer i]
+  #(let [local-time (-> (.getTime rs i)
+                        sql-time->local-time)
+         type-name  (.getColumnTypeName rs-meta i)
+         base-type  (presto-common/presto-type->base-type type-name)]
```
You should do the `type-name` and `base-type` and `isa?` stuff _once_, and then return an appropriate thunk function based on that. Otherwise we're potentially do that calculation millions of times... with the multimethod dispatch and the fact that `presto-type->base-type` does a linear comparison against a bunch of regexes), that overhead is definitely going add up. E.g. if you have a million-row result set with 3 TIME columns = 3 million calculations (instead of just doing 3, one for each column)

(That's the point of `read-column-thunk` -- return a minimal thunk function that can be used to fetch the value for a given column in the current row of the result set).

camsaul commented on modules/drivers/presto-jdbc/test/metabase/test/data/presto_jdbc.clj:
```diff
@@ -0,0 +1,135 @@
+  [& args]
+  (apply execute/sequentially-execute-sql! args))
+
+(remove-method load-data/load-data! :presto-jdbc)
```
Don't think you need to check this in

camsaul commented on modules/drivers/presto-jdbc/test/metabase/test/data/presto_jdbc.clj:
```diff
@@ -0,0 +1,135 @@
+(defmethod load-data/do-insert! :presto-jdbc
+  [driver spec table-identifier row-or-rows]
+  (let [statements (ddl/insert-rows-ddl-statements driver table-identifier row-or-rows)]
+    (with-open [conn (jdbc-spec->connection spec)]
```
How did this not trigger the reflection warning linter? You should add a `^java.sql.Connection` tag on `jdbc-spec->connection`. Otherwise `with-open` doesn't know that `conn` is an `AutoCloseable`

jeff303 replied:
Not sure why the reflection warner didn't complain, but I fixed them.

camsaul commented on modules/drivers/presto/test/metabase/test/data/presto.clj:
```diff
@@ -40,6 +42,7 @@
 
 (defn- field-base-type->dummy-value [field-type]
   ;; we need a dummy value for every base-type to make a properly typed SELECT statement
+  (println "field-type: " field-type)
```
Don't think you need this anymore

camsaul commented on modules/drivers/presto/test/metabase/test/data/presto.clj:
```diff
@@ -129,15 +132,15 @@
+     ((get-method tx/aggregate-column-info ::tx/test-extensions) driver ag-type))
 
-  ([driver ag-type field]
-   (merge
-    ((get-method tx/aggregate-column-info ::tx/test-extensions) driver ag-type field)
-    (when (= ag-type :sum)
-      {:base_type :type/BigInteger}))))
+    ([driver ag-type field]
+     (merge
+      ((get-method tx/aggregate-column-info ::tx/test-extensions) driver ag-type field)
+      (when (= ag-type :sum)
+        {:base_type :type/BigInteger}))))
```
Is it worth just removing this entirely?

jeff303 replied:
Yes, done.  It lives in `:presto-common` now.

camsaul commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -68,6 +68,18 @@
   additional keys there and need to access them."
   nil)
 
+(def ^:dynamic *parent-expr-type-info*
+  "Bound to the type info (i.e. result of calling (hx/type-info expr)) of a parent expression. Useful if a child
+  expression needs to vary its behavior according to the parent's type. Rebind as necessary when expanding complex
+  expressions.
+
+  Example: in Presto, checkins.date AT TIME ZONE 'America/Los_Angeles' is not allowed, because that column is of
+  type 'date', and the 'AT TIME ZONE' operator can only be invoked on types 'timestamp with time zone' and 'time
+  with time zone'. However, if a date bucketing expression is being expanded, the innermost ->honeysql form may
+  not have access to the original column's type info (since it's actually inside a SQL call to add or subtract dates,
+  etc."
+  nil)
+
```
It seems like this is going to break as soon as the expression ends up in a situation where the _grandparent_ has the type info rather than the parent. It's also extremely misleading to document this variable as being `Bound to the type info [...] of a parent expression` when it's actually only bound that way in one single very specific situation. It's also a little misleading in the sense that the `expr` that gets passed to `(date :week)` isn't really what I'd call the "parent expression" if it's getting wrapped and passed to other methods as `expr`... I don't expect `*parent-expr-type-info*` to mean "the type info of the arg you're passed before it got wrapped".

It seems sort of weird that you'd need to know the parent expression type information anyway in order to do something like apply `week-of-year` bucketing to a date column. I'm guessing you're doing it so the `:day-of-year` method sees it. Binding a dynamic variable in order to pass along type information about the expression we've wrapped seems like the wrong way to go about it. Why not just have the wrapped expression include type information?

Just make sure the `:week` implementation returns a HoneySQL form with type information. Arguably all the 
`sql.qp/date` truncation methods should be returning HoneySQL forms with type info. Any sort of truncation operation that returns another temporal type is returning _some_ type of value, maybe a `DATE` or `DATETIME` or `TIMESTAMP` or `TIME` or something else. We should pass that information along so we don't have to resort to weird dynamic variable hacks.

(That's the whole point of having the `hx/with-type-info` stuff in the first place -- so we can pass along information about what type a HoneySQL expression produces so we can do the appropriate thing in places where we need different functions/casts for different types without having to bind a bunch of dynamic variables that make the logic hard to follow.)

jeff303 replied:
Fair point.  Let me step back and explain the problem and maybe you can suggest a better solution.

Here is the problematic expression:
```
(hsql/call
 :date_add
 (literal "day")
 -1
 (hsql/call
  :date_trunc
  (literal "week")
  (hsql/call
   :date_add
   (literal "day")
   1
   (metabase.util.honeysql-extensions/with-type-info
    (hx/identifier :field "default" "checkins" "date")
    #:metabase.util.honeysql-extensions{:database-type "date"}))))
```

This is basically part of a `:day-of-week` truncation (which of course comes from `(defmethod date [:sql :week-of-year]...` in the QP itself).

Now, the current code in Presto JDBC (which you originally helped write) is basically using this `AtTimeZone` mechanism to add the `AT TIME ZONE` bit when the report timezone is defined.  The trouble is, this type of bucketing can very well be applied to both `date`, `timestamp`, and `timestamp with time zone` columns.   **Only** in the latter two cases, should this expression be added to the native query.  If we add it when the "underlying" (sorry, lack of a better term for it) column is a type `date` (as in the example above), then this will fail with:

```Type of value must be a time or timestamp with or without time zone (actual date)```

So, the task at hand is to **suppress** this extra expression in the case of date.  I attempted to do this poorly through the dynamic var.  Now, ripping that out as per your feedback... what next?  Should the `AtTimeZone` function walk the entire expression tree searching for the "underlying" base type (which, as can be seen, is available in some fashion).

jeff303 replied:
OK, I think I figured out how to deal with it properly: basically by switching from an exclusion list to an inclusion one.  In the `AtTimeZone` handling, I will only set the expression if the type is KNOWN to be a valid one (`time`, `time with time zone`, `timestamp`, or `timestamp with time zone`).  That way, the innermost call (i.e. the column itself) will be handled correctly, and the wrapping function calls don't need to care about it (and the Presto query engine presumably should just do the right thing in that case).

camsaul commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -109,6 +121,17 @@
   [driver]
   :%now)
 
+(defmulti current-datetime-base-type
+  "Field :base-type that is returned by the DB specific \"current datetime\" HSQL form
+  (i.e. `current-datetime-honeysql-form`).  Defaults to `:type/DateTime`."
+  {:arglists '([driver])}
+  driver/dispatch-on-initialized-driver
+  :hierarchy #'driver/hierarchy)
+
+(defmethod current-datetime-base-type :default
+  [_]
+  :type/DateTime)
```
1. Seeing how this is only used in tests, it should be a test method rather than a driver method itself. 3rd party driver authors or future driver authors at MB are going to see this and not be sure if they need to implement it or why they'd want to do so
2. Why not just update the dox for `current-datetime-honeysql-form` and say that it should preferably return a HoneySQL form with type information? And have `:presto-jdbc` do this. This is arguably something they should already be doing anyway, for the same reason all the other temporal-related stuff should be doing it -- so we can avoid unneeded casts/use them when needed/use appropriate functions for databases that don't allow implicit conversions between different temporal types.  If the form has type info, the tests could convert the database type to a metabase base type using `database-type->metabase-type`, otherwise fall back to `:type/DateTime`
3. New driver multimethods, if you do add them, should have `:added "0.40.0"` metadata so we can keep track of when we added stuff so we can inform 3rd-party driver authors

camsaul replied:
Since this information could potentially come back as part of the type information on the HoneySQL form returned by `current-datetime-honeysql-form`, if we _were_ to add this method, we could potentially have conflicts between what the two methods returned. e.g. maybe my driver has

```clj
(defmethod current-datetime-honeysql-form :my-driver
  [_]
  (hx/with-database-type-info :%now "date"))

(defmethod current-datetime-base-type :my-driver
  [_]
  :type/DateTimeWithTZ)
```

That's clearly _wrong_, and it's not clear which one should be considered correct. If we're trying to make sure we cast `current-datetime-honeysql-form` correctly, do we treat it as a `date`, or as whatever `(database-type->honeysql-type :type/DateTimeWithTZ)` returns?

The fact that this method would introduce a second source of truth should be a red flag that we might not actually need it

jeff303 replied:
OK, good points.  I got rid of the multimethod and instead just changed `current-datetime-honeysql-form` to attach the database type to the form (for Presto JDBC) and changed the test to check for that.

camsaul commented on test/metabase/query_processor_test/date_bucketing_test.clj:
```diff
@@ -1029,6 +1032,7 @@
              (mt/formatted-rows [int]
                (mt/run-mbql-query checkins
                  {:aggregation [[:count]]
+                  ;; TODO: somehow this is getting turned into timestamp '2019-01-16 00:00 UTC'
```
Is this TODO still true? If it is, it might be worth clarifying that it's getting turned into something for Presto JDBC and not generally.

jeff303 replied:
Nope, removed now.

camsaul commented on test/metabase/util/date_2_test.clj:
```diff
@@ -586,3 +586,8 @@
+
+(deftest standard-offset-test
+  (testing "standard-offset works correctly"
+    (is (= (t/zone-offset "-06:00") (u.date.common/standard-offset (t/zone-id "America/Chicago"))))
```
You should add some tests that use `mt/with-clock` to switch the clock to something definitely in DST and something definitely not in DST to make sure it returns the same offset regardless of the system clock date

camsaul reviewed (changes requested):
Overall this looks good but I left a bunch of notes on relative small things. I'll try to summarize here:

-  Driver methods should pass along the `driver` they are called with when calling other methods rather than hardcoding the driver for which the method is defined. For example if I call a `:presto-common` method `a` with `:presto`, and `a` calls method `b` with `:presto-common`, then it's the `:presto-common` impl of `b` that gets used and not the `:presto` impl like one would expect.
- Don't special case results in tests for different drivers when you can use `mt/formatted-rows`
- Add a note explaining why you're skipping a test for a driver
- `read-column-thunk` should do any calculations that can be done on `ResultSetMetaData` and column number alone once and return a thunk rather than doing them for every single value fetched from the DB. This can be the difference between that calculation being done 3 times and 3 million times -- and since mapping presto DB type to Metabase base type is a linear search thru a list of regexes we're actually talking about maybe 30 regex comparisons vs 30 million
- Don't forget to type tag your stuff. `(set! *warn-on-reflection* true)` when you start up your REPL and then fix the warnings
- have your impls of `sql.qp/date` pass along type info rather than defining a new dynamic variable that's used in exactly one place and only for a single driver. The whole point of the the HoneySQL type info stuff is to avoid needing dynamic variables like that
-  We should change the docstring for `current-datetime-honeysql-form` to say that you should return a form with type info (which we should do for other reasons anyway) and then use that info to determine the base type it returns in the date bucketing test rather than adding a completely new driver method.

jeff303 committed:
PR feedback

Moving CircleCI env vars as per Cam's comment

jeff303 committed:
PR feedback

Enhancing additional-unit-filtering-tests to cover setting report TZs as well

Removing current-datetime-base-type multimethod

Fixing thunk implementations to be much more efficient

jeff303 committed:
Fix trailing blank line

jeff303 committed:
Switch away from LocalTime/ofInstant method since it's not available in Java 8

jeff303 committed:
Fix linter errors

jeff303 committed:
Fix ns ordering, don't invoke multimethod with null value

jeff303 committed:
Remove trace log

jeff303 committed:
Incorporating report-timezone into unix-timestamp->honeysql methods after all

Changing Presto so that it can recognize :type/DateTimeWithZoneOffset (so that the changes to the office checkins dataset work there)

jeff303 committed:
More tweaking timezone code for Presto JDBC

Adding long comment documenting one of the failures

jeff303 committed:
Switching sql.qp/date implementation to use Presto `date` function instead of `date_trunc` with 'day' to avoid timezone conversion issues

Adding impl of ->prepared-substitution for LocalDate that converts to midnight at the start of that day in the report timezone

Making sync only operate on the given schema connection property, if specified, and adding test for that

jeff303 committed:
Remove ->prepared-substitution for LocalDate since that introduced test failures

dpsutton commented on modules/drivers/presto-common/src/metabase/driver/presto_common.clj:
```diff
@@ -0,0 +1,203 @@
+;; Presto mod is a function like mod(x, y) rather than an operator like x mod y
+(defmethod hformat/fn-handler (u/qualified-name ::mod)
+  [_ x y]
+  (format "mod(%s, %s)" (hformat/to-sql x) (hformat/to-sql y)))
```
can you not do `(hsql/call :mod x y)`?

jeff303 replied:
Apparently not.  The `hsql/call` ends up calling **this** (`fn-handler`) to figure out the "form" the function invocation could take.  There are a couple other drivers that do this exact same thing (I copied the method from there; see Oracle and BigQuery).  To be honest, though, I'm unsure where the `x mod y` (i.e. the default) form is even defined.  I tried to find it and couldn't.

dpsutton commented on modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj:
```diff
@@ -0,0 +1,489 @@
+  (let [report-zone (qp.timezone/report-timezone-id-if-supported :presto-jdbc)]
+    (hsql/call :from_unixtime expr (hx/literal (or report-zone "UTC")))))
+
+(defmethod sql.qp/unix-timestamp->honeysql [:presto-jdbc :milliseconds]
```
I suppose this is fine as well, but other implementations are fine with just dividing by 1000 and doing a seconds conversion. Seems either is fine.

jeff303 replied:
Yeah, I figure that since Presto does support millisecond precision, might as well use it even though it's slightly complicated.

dpsutton commented on modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj:
```diff
@@ -0,0 +1,489 @@
+        (.close conn)
+        (throw e)))))
+
+#_(defmethod sql-jdbc.execute/set-parameter [:presto-jdbc OffsetDateTime]
```
just confirming, are these meant to be commented out?

jeff303 replied:
I had a bit of back and forth with Cam on this, on a different thread (don't know how to link directly to other GitHub PR comment threads).  But in essence, this is theoretically not needed anymore because I have overridden the `->prepared-substitution` method for all of these temporal types, which then go on to set the "values" as strings.  However, if there are any other code paths that end up calling `set-parameter`, then those will definitely fail with an error, because the Presto JDBC driver doesn't support the override that ends up being called. 

The fact that they're commented out, but the tests pass, suggests strongly that no such code path exists, but I was considering adding them back in just out of sheer paranoia.  I think for now, I will just remove and then we can add them if we have an identified bug in the future.

dpsutton replied:
cool. yeah i vote remove them as well

dpsutton commented on modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj:
```diff
@@ -0,0 +1,489 @@
+  (let [type-name  (.getColumnTypeName rs-meta i)
+        base-type  (presto-common/presto-type->base-type type-name)
+        with-tz?   (isa? base-type :type/TimeWithTZ)]
+    #(let [local-time (-> (.getTime rs i)
```
small nit but not a fan of returning a thunk like this. very easy to miss that it's a function rather than just a regular expression. Seems like all the others use `(fn []` and some even have names

jeff303 replied:
Fixed.

dpsutton commented on src/metabase/sync/sync_metadata.clj:
```diff
@@ -49,7 +54,14 @@
+  synced if the table's database's driver supports the `:foreign-keys` feature."
+  [tbl :- i/TableInstance]
+  (sync-fields/sync-fields-for-table! tbl)
+  (let [drv (-> tbl table/database driver.u/database->driver)]
```
surprising that a bug like this has existing for so long

dpsutton replied:
ah, there's a check for this below. this is just an optimization it appears?

jeff303 replied:
Unsure what you mean by "below", here?  But I think that without checking whether the driver supports FKs, the `sync-fks-for-table!` will get called, and may fail (or simply do nothing, depending on how the JDBC driver behavior works).  I don't think it breaks sync or anything when that happens, since it will just capture that error status and continue doing the other tasks.  But it's a bit of an optimization, I suppose, so that it will only be attempted if it's thoeretically supported.

dpsutton replied:
https://github.com/metabase/metabase/blob/master/src/metabase/sync/fetch_metadata.clj#L24
This is what they eventually call and it's a no-op when the driver doesn't support it it seems. By below i meant further in the call stack.

jeff303 replied:
Ah, not sure how I missed that!  I'm also unsure of the circumstances in which I was seeing the exception from the JDBC driver pertaining to that feature not being supported.  Thanks; I have backed out the change to this file.

jeff303 committed:
PR feedback from Dan

jeff303 committed:
Remove unused binding

jeff303 committed:
PR feedback

Reverting unnecessary change

jeff303 commented:
Final merge should not happen until #17028 is merged, at which point, the existing Presto driver can be marked as superseded by the new one.  Also, this PR will be retargeted to 40.

jeff303 committed:
Incorporate driver deprecation for Presto and Presto JDBC drivers

jeff303 committed:
Try to fix "use legacy" link via selectEngine fn

jeff303 committed:
Try to fix "use legacy" link via selectEngine fn

jeff303 committed:
Revert "Try to fix "use legacy" link via selectEngine fn"

This reverts commit 4770abae

jeff303 committed:
Revert "Try to fix "use legacy" link via selectEngine fn"

This reverts commit 4770abae

jeff303 committed:
Switch doc link to 01-managing-databases

jeff303 committed:
Switch doc link to 01-managing-databases

jeff303 committed:
Fix linter error

jeff303 committed:
Do not require env var anymore, now that deprecation is in place

jeff303 committed:
Handle airports catalog

jeff303 committed:
Include diagnostic info
